### DISCUSSION TITLE
Streaming Bluetooth audio to internal DAC and external I2S DAC at the same time

### DISCUSSION THREAD

**Discussion by MrTwisterFister69** [pschatzmann/ESP32-A2DP#318] [Q&A] [answered]
Hello,
Great Work!
I would like to build a bluetooth speaker, which also has sound control lights. That's why i want to stream the audio to an external I2S DAC for the amp and to the internal DAC of the ESP 23 for the lights.
But I am not sure how to combine both correctly.

**Comment by pschatzmann** [comment]
I am not sure what you mean with lights ? 
Should the volume control the brightness of an LED ?

I don't think that this is an issue but rather a discussion, so I convert it...

**Comment by MrTwisterFister69** [comment]
Yes that's exactly what I want to do!
Sorry for posting my question as an issue, I am new to Github.

**Comment by pschatzmann** [comment, marked as answer]
You would use the callback described in 'Accessing the Sink Data Stream with Callbacks'. From that you would analyse the data by calculating a max value (or avg) over a range to use as volume. Then you would translate the volume into into a range between 0 and 254 and use an analogWrite() to output to the LED.

If you want to minimize the lines of code you can also use the [AudioTools](https://pschatzmann.github.io/arduino-audio-tools) library and just write to the [VolumePrint audio sink](https://pschatzmann.github.io/arduino-audio-tools/html/classaudio__tools_1_1_volume_print.html) and get the volume from there...

I suggest that you have a look at the following examples to get you started:
- https://github.com/pschatzmann/ESP32-A2DP/blob/main/examples/bt_music_receiver_datacallback/bt_music_receiver_datacallback.ino
- https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-basic-api/basic-a2dp-fft/basic-a2dp-fft.ino

**Comment by mrhempman69** [comment]
But to be clear, it's not possible to output to the internal DAC and an external DAC (or, an external one over i2s), because i2s requires the internal DAC pins as part of its protocol... right?

**Comment by pschatzmann** [comment]
If you mean the Analog Output via I2S, sure you can: the only restriction is that the internal dac pins are fixed and must use port 0 for it. 
So it will be port 1 for the additional i2s input or output.

But I would not use the audio output, but just some regular analogWrite() for this purpose.

**Reply by ThomasCurnow** [reply]
Hi Phil,
Just wondering how I can use BluetoothA2DPSink to write to these two streams at the same time? I want to be able to output the audio to I2S which would use port 1, but I also want to simultaneously output it to an AnalogAudioStream, presumably on port 0. 

For this application, I cannot use the method above where we calculate the "loudness" and then use analogWrite(), it must output the audio from the internal DAC. Thank you.

**Reply by pschatzmann** [reply]
I don't see any reason why this should not work: just use a MultiOutput!

**Reply by ThomasCurnow** [reply]
MultiOutput proved to be the perfect solution to this problem. Thank you
